### DISCUSSION TITLE
How to translate this C# preprocessor to PascalABC.NET?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3004] [Q&A] [answered]
```
#if WIN32
    [STAThreadAttribute()]
#endif
```

**Comment by SunSerega** [comment]
`[STAThread]` is only useful in C# when applied to the main function, which is not possible in Pascal. Otherwise, it's better to launch a new thread during execution and configure it manually, then you can just check platform bit-ness in regular code.

In the first place, why would you want to switch thread to STA only on 32-bit platforms? And how is that supposed to work in .Net, where platform bit-ness is usually decided when .exe starts executing, and compiler directives are processed before creating an .exe file?

**Reply by ghost** [reply]
The library is a cross platform GUI toolkit written in C with bindings for .NET, and I'm just translating the C# code to PascalABC.NET. The preprocessor makes sense, since only Windows has this COM Apartments stuff.

**Reply by SunSerega** [reply, marked as answer]
I see, so "Win32" was also applied to Win64...

Well, either way, that's a very bad solution, it would require you to compile a different .exe for every platform.
Much better solution is launching new thread manually and doing a runtime check:
```
var thr := new System.Threading.Thread(()->
begin
  //Launch window and message loop here
end);

if System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows) then
  // System.PlatformNotSupportedException on non-windows platforms
  thr.SetApartmentState(System.Threading.ApartmentState.STA);

thr.Start;
```

**Reply by SunSerega** [reply]
Btw there are many ways to check for OS type, this one is one of the oldest.

And while this is a slower code, it might make sense to do something like:
```
try
  thr.SetApartmentState(System.Threading.ApartmentState.STA);
except
  on System.PlatformNotSupportedException do ;
end;
```
Without checking for the platform first. Since this would skip setting the apartment state only when it's not supported on the current platform.
